solr/core/src/java/org/apache/solr/request/macro/MacroExpander.java

```diff
@@ -188,7 +189,7 @@ else if (failOnMissingParams) {
 
       } catch (SyntaxError syntaxError) {
         // append the part we would have skipped
-        sb.append( val.substring(matchedStart, start) );
+        sb.append(val, matchedStart, start);
```

Comment:
observation: `foo.append(bar.substring(x,y));` is also found in other places in the code base, not sure how it might work implementation wise but it would be lovely if tooling would flag up that there's a more efficient alternative

Reply:
IntelliJ pointed this out to me :-)  It didn't know about the val.substring(start) (implied val.length()) so I did those manually.  Yes, it'd be nice if everywhere we could do this in one go.  Feel free to file an issue to do so if you have time.  IntelliJ can make quick work of that.